### PR TITLE
FSF ROS

### PR DESCRIPTION
**Pull request by igorljubuncic**
This is the updated FSF ROS guide.

The YAML shown on the template page builds correctly, and the snap runs.

The description primarily highlights the benefits of snaps to the robotics audience. I've made the changes after consulting with Kyle.

I've also updated the test section to reflect the correct invocation of the snap, as the snap name and the application name are different.

### CODE REVIEW

**Comment by kyrofa** [comment]
> I've also updated the test section to reflect the correct invocation of the snap, as the snap name and the application name are different.

It definitely doesn't need to be different. If the pattern we have so far is the same, I see no problem with following that here.

**Comment by igorljubuncic** [comment]
Kyle, I will amend the FSF to follow the same usage convention we have in the other guides.

File: webapp/first_snap/content/ros/build.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    - action: You will find a .snap file in the same directory that you ran the snapcraft command. You can inspect the contents to ensure it contains all of the application's assets
+      command: unsquashfs -l *.snap
+    - action: Install the snap
+      command: sudo snap install --devmode --dangerous *.snap
+    - action: List your installed snaps to confirm
+      command: snap list
+    - action: Run the snap
+      command: test-ros-talker-listener-{name} -h
+    - warning: |
+        Make sure the name matches what has been used previously (i.e <code>test-ros-talker-listener-{name}</code>)
```
Comment (kyrofa) [inline webapp/first_snap/content/ros/build.yaml]: e.g.?

File: webapp/first_snap/content/ros/build.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  manual:
+    - action: Return to the root directory of the project containing your snapcraft.yaml and run the snapcraft Docker image
+      command: docker run -v $PWD:$PWD -w $PWD snapcore/snapcraft snapcraft
```
Comment (kyrofa) [inline webapp/first_snap/content/ros/build.yaml]: As in ROS2, this won't work.

File: webapp/first_snap/content/ros/build.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  auto:
+    - action: Return to the root directory of the project containing your snapcraft.yaml and run snapcraft
+      command: snapcraft
+    - action: You will find a .snap file in the same directory that you ran the snapcraft command. You can inspect the contents to ensure it contains all of the application's assets
+      command: |
+        brew install squashfs
+        unsquashfs -l *.snap
+  manual:
+    - action: Return to the root directory of the project containing your snapcraft.yaml and run the snapcraft Docker image
+      command: docker run -v $PWD:$PWD -w $PWD snapcore/snapcraft snapcraft
```
Comment (kyrofa) [inline webapp/first_snap/content/ros/build.yaml]: Ditto.

File: webapp/first_snap/content/ros/build.yaml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+windows:
+  auto:
+    - action: Run Windows Subsystem for Linux from the Windows Start menu
+    - action: Return to the root directory of the project containing your snapcraft.yaml and run snapcraft
+      command: snapcraft
+    - action: You will find a .snap file in the same directory that you ran the snapcraft command. You can inspect the contents to ensure it contains all of the application's assets
+      command: unsquashfs -l *.snap
+  manual:
+    - action: Return to the root directory of the project containing your snapcraft.yaml and run the snapcraft Docker image
+      command: docker run -v $PWD:$PWD -w $PWD snapcore/snapcraft snapcraft
```
Comment (kyrofa) [inline webapp/first_snap/content/ros/build.yaml]: Again.

File: templates/home/_fsf_ros.html [hunk 5]
```diff
@@ -2,15 +2,40 @@
...
+      <li>Bundle all the runtime requirements, including the exact version of ROS and system libraries you need.</li>
       <li>Expand the distributions supported beyond just Ubuntu.</li>
-      <li>Directly control the delivery of application updates.</li>
+      <li>Directly and reliably control the delivery of application updates using existing infrastructure.</li>
       <li>Extremely simple creation of daemons.</li>
     </ul>
   </div>
 
+  <div class='col-6'>
+    <h4>How ros-talker-listener defines snapcraft.yaml:</h4>
```
[the remark this answers is not in the document]
Comment (igorljubuncic) [reply templates/home/_fsf_ros.html]: Agreed!

File: templates/home/_fsf_ros.html [hunk 6]
```diff
@@ -2,15 +2,40 @@
...
+  <b>ros-tutorials</b>:
+    <b>plugin</b>: catkin
+    <b>source</b>: https://github.com/ros/ros_tutorials.git
+    <b>source-branch</b>: melodic-devel
+    <b>source-space</b>: roscpp_tutorials/
+
+<b>apps</b>:
+  <b>ros-talker-listener</b>:
+    <b>command</b>: roslaunch roscpp_tutorials \
+    talker_listener.launch
```
[the remark this answers is not in the document]
Comment (igorljubuncic) [reply templates/home/_fsf_ros.html]: Agreed.

File: webapp/first_snap/content/ros/package.yaml [hunk 7]
```diff
@@ -0,0 +1,59 @@
+name: test-ros-talker-listener-{name}
+download: git clone https://github.com/snapcraft-docs/ros-talker-listener
```
[the remark this answers is not in the document]
Comment (igorljubuncic) [reply webapp/first_snap/content/ros/package.yaml]: I would suggest the following:

- We test with the yaml provided with this PR.
- Once you are @evandandrea are happy, we then commit the yaml to the repo and overwrite the existing one there.

File: webapp/first_snap/content/ros/package.yaml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+createDir: |
+  cd ros-talker-listener
+  $EDITOR snapcraft.yaml
```
[the remark this answers is not in the document]
Comment (igorljubuncic) [reply webapp/first_snap/content/ros/package.yaml]: For commonality, and in line with the other examples we have, I would keep it in the top-level directory.

File: webapp/first_snap/content/ros/package.yaml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+metadata: |
+  name: ros-talker-listener
+  version: '0.1'
+  summary: ROS Talker/Listener Example
+  description: |
+    This example launches a ROS2 talker and listener.
```
[the remark this answers is not in the document]
Comment (igorljubuncic) [reply webapp/first_snap/content/ros/package.yaml]: Agreed.

File: webapp/first_snap/content/ros/package.yaml [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        Parts define what sources are needed to assemble your app. Parts can be anything: programs, libraries, or other needed assets. The <code>catkin</code> plugin is useful when building ROS application parts.
+    - code: |
+        parts:
+          ros-tutorials:
+            plugin: catkin
+            source: https://github.com/ros/ros_tutorials.git
+            source-branch: melodic-devel
+            source-space: roscpp_tutorials/
+    - text: |
+        You can also use the optional <code>rosdistro</code> keyword to change the ROS distro required by the system. By default, <code>Melodic</code> is used with the <code>core18</code> base. The command <code>snapcraft help catkin</code> gives more information on how this plugin works.
```
[the remark this answers is not in the document]
Comment (igorljubuncic) [reply webapp/first_snap/content/ros/package.yaml]: Agreed, thanks.

File: webapp/first_snap/content/ros/package.yaml [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+  apps:
+    - text: Apps are the commands you want to expose to users and any background services your application provides.
+    - code: |
+        apps:
+          ros-talker-listener:
+            command: roslaunch roscpp_tutorials talker_listener.launch
+    - warning: |
+        <span class="p-notification__status">Snap names are globally unique.</span> You should change any instance of <code>ros-talker-listener</code> to <code>test-ros-talker-listener-{name}</code> (where <code>{name}</code> is your name) before continuing to avoid conflicting with the official ros-talker-listener snap.
```
[the remark this answers is not in the document]
Comment (igorljubuncic) [reply webapp/first_snap/content/ros/package.yaml]: This warning exists in all the language guides. For consistency, I believe it should remain there so all guides have the same flow. @evandandrea Should we keep the second name warning text box in place?

File: webapp/first_snap/content/ros/package.yaml [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+  parts:
+    - text: |
+        Parts define what sources are needed to assemble your app. Parts can be anything: programs, libraries, or other needed assets. The <code>catkin</code> plugin is useful when building ROS application parts.
+    - code: |
+        parts:
+          ros-tutorials:
+            plugin: catkin
+            source: https://github.com/ros/ros_tutorials.git
+            source-branch: melodic-devel
+            source-space: roscpp_tutorials/
```
[the remark this answers is not in the document]
Comment (igorljubuncic) [reply webapp/first_snap/content/ros/package.yaml]: Agreed.

File: webapp/first_snap/content/ros/test.yaml [hunk 13]
```diff
@@ -0,0 +1,16 @@
+macos:
+  - action: Create a Linux virtual machine for testing snaps
+    command: multipass launch -n testvm
+  - action: Return to the directory containing the .snap file and copy it into the virtual machine
+    command: 'multipass copy-files ros-talker-listener*.snap testvm:'
+  - action: Connect to the virtual machine
+    command: multipass shell testvm
+  - action: Install the snap inside the virtual machine
+    command: sudo snap install --dangerous ros-talker-listener*.snap
```
[the remark this answers is not in the document]
Comment (igorljubuncic) [reply webapp/first_snap/content/ros/test.yaml]: @evandandrea This affects all language guides - can we do a single high-level template change, or upload a revision for each guide separately?

File: webapp/first_snap/content/ros/build.yaml [hunk 14]
```diff
@@ -0,0 +1,54 @@
+linux:
+  auto:
+    - action: Return to the root directory of the project containing your snapcraft.yaml and run snapcraft
```
[the remark this answers is not in the document]
Comment (igorljubuncic) [reply webapp/first_snap/content/ros/build.yaml]: For consistency, we should keep it snapcraft.yaml.

File: webapp/first_snap/content/ros/build.yaml (hunk 1 above)
Comment (igorljubuncic) [reply webapp/first_snap/content/ros/build.yaml]: I think this makes sense, as it means - that is - so this is very specific.

File: webapp/first_snap/content/ros/build.yaml (hunk 2 above)
Comment (igorljubuncic) [reply webapp/first_snap/content/ros/build.yaml]: If you choose macOS in the guide, the `manual` option is not shown, so the docker usage is not presented to the user. @evandandrea do we want to do a site-wide cleanup of the manual configuration option?

Comment (evandandrea) [reply webapp/first_snap/content/ros/build.yaml]: Ignore the Docker stuff. I'll work on a PR to drop that from all the first snap flows.

File: webapp/first_snap/content/ros/build.yaml (hunk 3 above)
Comment (evandandrea) [reply webapp/first_snap/content/ros/build.yaml]: As above, ignore.

File: webapp/first_snap/content/ros/build.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (evandandrea) [reply webapp/first_snap/content/ros/build.yaml]: Yes, we can fix up the repo when landing this. The snapcraft.yaml should be in the root, matching the other first snap flow examples.

File: webapp/first_snap/content/ros/package.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (evandandrea) [reply webapp/first_snap/content/ros/package.yaml]: Yes, we can fix up the repo when landing this. The snapcraft.yaml should be in the root, matching the other first snap flow examples.

File: webapp/first_snap/content/ros/package.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (evandandrea) [reply webapp/first_snap/content/ros/package.yaml]: +1

File: package.json [hunk 15]
```diff
@@ -46,7 +46,6 @@
     "babel-preset-env": "^1.6.1",
     "babel-preset-react": "^6.24.1",
     "clipboard": "^2.0.0",
-    "cssnano": "^3.10.0",
```
Comment (evandandrea) [inline package.json]: It looks like you need to rebase on master. Your diff shouldn't show changes outside the pages you've explicitly modified.

Comment (igorljubuncic) [reply package.json]: Not sure why.

* I always git fetch before making changes.
* I only git added the relevant files (template and yaml), definitely not this.

Comment (kyrofa) [reply package.json]: A `git fetch` only updates the remote branches, not locals (which are tracking remotes). After a fetch you can try switching to `master` and then run `git reset --hard origin/master` for example to bring it up to date. Or just switch to `master` and run `git pull` which does a fetch and merge for you.

File: webapp/first_snap/content/ros/package.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (kyrofa) [reply webapp/first_snap/content/ros/package.yaml]: You marked this resolved, but then didn't fix it?

Comment (kyrofa) [inline webapp/first_snap/content/ros/package.yaml]: This still says ROS2. My previous comment saying that was marked resolved.

**Comment by kyrofa** [comment]
> I'd like @kyrofa to have another look over before we land this.

Happy to once the diff is more manageable.

**Comment by igorljubuncic** [comment]
> > I'd like @kyrofa to have another look over before we land this.
> 
> Happy to once the diff is more manageable.

This is all here:
https://github.com/canonical-websites/snapcraft.io/pull/1641/commits/35d6891e07a5b3fbb3d974fa4af56cedfe5d006d

File: webapp/first_snap/content/ros/test.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (evandandrea) [reply webapp/first_snap/content/ros/test.yaml]: Wow, how did we miss this? Make the change here and I promise to make the change in all other guides now.

[the remark this answers is not in the document]
Comment (evandandrea) [reply webapp/first_snap/content/ros/test.yaml]: As promised: https://github.com/canonical-websites/snapcraft.io/pull/1729

File: webapp/first_snap/content/ros/build.yaml (hunk 14 above)
[the remark this answers is not in the document]
Comment (kyrofa) [reply webapp/first_snap/content/ros/build.yaml]: Done, I updated the repo.

File: webapp/first_snap/content/ros/package.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (kyrofa) [reply webapp/first_snap/content/ros/package.yaml]: Done and done, the repo is tested and updated.

File: webapp/first_snap/content/ros/package.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (kyrofa) [reply webapp/first_snap/content/ros/package.yaml]: Fixed.

**Comment by kyrofa** [comment]
This looks good to me except for [one unaddressed comment](https://github.com/canonical-websites/snapcraft.io/pull/1641/files#r262676281) by @evandandrea.

File: webapp/first_snap/content/ros/package.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (evandandrea) [reply webapp/first_snap/content/ros/package.yaml]: Why do you think it's out of place, @kyrofa? If `ros-talker-listener` under `apps:` doesn't match the `name:`, they'll end up with a namespaced command. Remember they'll be pushing this to the store, so it necessitates choosing a different name than `ros-talker-listener` (`ros-talker-listener-kyrofa`).

File: webapp/first_snap/content/ros/test.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (evandandrea) [reply webapp/first_snap/content/ros/test.yaml]: To be clear, this PR still needs to be updated to add `--devmode` here.

File: webapp/first_snap/content/ros/package.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (kyrofa) [reply webapp/first_snap/content/ros/package.yaml]: It's the exact same warning [they already saw](https://github.com/canonical-websites/snapcraft.io/pull/1641/files#diff-30766645475fe82a5698f78df4142ad4R18), so it looks like a mistake. This seems like a more proper place to talk about exactly what you just said: the fact that, if the app name is the same as the snap name, it's not namespaced. Not repeat that snap names need to be unique.

[the remark this answers is not in the document]
Comment (evandandrea) [reply webapp/first_snap/content/ros/package.yaml]: How about:
>"As before, remember to change this instance of ros-talker-listener to test-ros-talker-listener-{name} (where {name} is your name)."

I think if we have no message at all, people won't remember to do it. Granted, it's not the worst thing if their command is namespaced, so if you feel strongly about this say so and we can drop the warning under `apps:`.

Worth noting that the next Design iteration of the first snap flow will let you pick your name with an input box and generate the snapraft.yaml for you. /cc @hughesjordan @therealjuan

[the remark this answers is not in the document]
Comment (kyrofa) [reply webapp/first_snap/content/ros/package.yaml]: Sure, I don't have super strong feelings here, I just wanted to offer some fresh eyeballs and have the discussion, thanks!

[the remark this answers is not in the document]
Comment (evandandrea) [reply webapp/first_snap/content/ros/package.yaml]: Understood, thanks.

@igorljubuncic if you change the text of the second warning to "As before, remember to change this instance of ros-talker-listener to test-ros-talker-listener-{name} (where {name} is your name)." I'll do the same for the other guides.

File: webapp/first_snap/content/ros/build.yaml [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+      command: snapcraft
+    - action: You will find a .snap file in the same directory that you ran the snapcraft command. You can inspect the contents to ensure it contains all of the application's assets
+      command: unsquashfs -l *.snap
+    - action: Install the snap
+      command: sudo snap install --devmode --dangerous *.snap
+    - action: List your installed snaps to confirm
+      command: snap list
+    - action: Run the snap
+      command: test-ros-talker-listener-{name} -h
```
Comment (evandandrea) [inline webapp/first_snap/content/ros/build.yaml]: Does `test-ros-talker-listener-{name} -h` work?

**Comment by igorljubuncic** [comment]
Yes, it prints the output of the options you need:

```
snap run ros-talker-listener -h
Usage: roslaunch [options] [package] <filename> [arg_name:=value...]
       roslaunch [options] <filename> [<filename>...] [arg_name:=value...]

If <filename> is a single dash ('-'), launch XML is read from standard input.

Options:
  -h, --help            show this help message and exit
  --files               Print list files loaded by launch file, including
                        launch file itself
  --args=NODE_NAME      Print command-line arguments for node
  --nodes               Print list of node names in launch file
  --find-node=NODE_NAME
...
```

File: templates/partials/fsf_language.html [hunk 17]
```diff
@@ -66,7 +66,7 @@
       </header>
```
Comment (evandandrea) [inline templates/partials/fsf_language.html]: You're missing the change for `/first-snap/ros`.

**Pull request merged**